### PR TITLE
Handle query parameters in FileMiddleware redirects

### PR DESCRIPTION
**Pull request by Captain-Kirkie**
Correctly handle query parameters when using the redirect functionality in `FileMiddleware`.

### CODE REVIEW

**Comment by Captain-Kirkie** [comment]
Issue [Here](https://github.com/vapor/vapor/issues/3076)

File: Sources/Vapor/Middleware/FileMiddleware.swift [hunk 1]
```diff
@@ -164,3 +165,16 @@ fileprivate extension String {
         return newPath
     }
 }
+
+
+extension Request {
+    public func getQueryString() -> String {
```
Comment (0xTim) [inline Sources/Vapor/Middleware/FileMiddleware.swift:171]: You can just use `req.url.query` instead

**Review by 0xTim — CHANGES_REQUESTED** [review]
Thanks for the PR! One change then we can merge

File: Sources/Vapor/Middleware/FileMiddleware.swift [hunk 2]
```diff
@@ -62,8 +62,9 @@ public final class FileMiddleware: Middleware {
             guard absPath.hasSuffix("/") else {
                 switch directoryAction.kind {
                 case .redirect:
+                    let queryStr = request.getQueryString()
                     return request.eventLoop.future(
-                        request.redirect(to: request.url.path + "/", redirectType: .permanent)
+                        request.redirect(to: request.url.path + "/" + queryStr, redirectType: .permanent)
```
Comment (gwynne) [inline Sources/Vapor/Middleware/FileMiddleware.swift:67]: Why not redirect the entire URL (with the modified path) so as to also preserve authority parameters (username, password) and other components such as the fragment (text following a `#`)?

Comment (0xTim) [reply Sources/Vapor/Middleware/FileMiddleware.swift:67]: That's a better solution

Not that fragment would be passed back to the server but the auth stuff would be

Comment (Captain-Kirkie) [reply Sources/Vapor/Middleware/FileMiddleware.swift]: That is more elegant. I have updated the PR to use the entire URL.

File: Sources/Vapor/Middleware/FileMiddleware.swift [hunk 3]
```diff
@@ -62,8 +62,9 @@ public final class FileMiddleware: Middleware {
             guard absPath.hasSuffix("/") else {
                 switch directoryAction.kind {
                 case .redirect:
+                    let redirectLocation = request.url.string.replacingOccurrences(of: request.url.path, with: "\(request.url.path)/")
                     return request.eventLoop.future(
-                        request.redirect(to: request.url.path + "/", redirectType: .permanent)
+                        request.redirect(to: redirectLocation, redirectType: .permanent)
```
Comment (gwynne) [inline Sources/Vapor/Middleware/FileMiddleware.swift:67]: This is unsafe if the path occurs multiple times in the URL (which is perfectly possible). Try something like this, maybe:
```suggestion
                    var redirectUrl = request.url
                    redirectUrl.path += "/"
                    return request.eventLoop.future(request.redirect(to: redirectUrl.string, redirectType: .permanent))
```

**Review by gwynne — CHANGES_REQUESTED** [review]

File: Sources/Vapor/Middleware/FileMiddleware.swift (hunk 3 above)
Comment (Captain-Kirkie) [reply Sources/Vapor/Middleware/FileMiddleware.swift]: whoops, good catch. Updated.

**Review by 0xTim — APPROVED** [review]
Thanks!

**Pull request merged**